### DISCUSSION TITLE
No layer count nor estimate shows while printing

### DISCUSSION THREAD

**Discussion by sn4k3** [mainsail-crew/mainsail#1863] [Help] [answered]
I'm using basic gcode with mSLA, but I followed all PrusaSlicer recommendations for notes, still I can't have layer count nor estimates to show.

I use:

- SET_PRINT_STATS_INFO
- M73 
- ; total layers count

Progress after 10 layers:

![image](https://github.com/mainsail-crew/mainsail/assets/113281/ccaacb76-aa05-415a-8789-9765aad55674)

GCode:

```
;Generated by UVtools v4.3.3 64-bits @ 29/04/2024 04:03:18
;FileName:Angel_Tablet_Full_Model_Resin_0.1mm_04h47m.zip
;MachineName:Uniz IBEE
;PrintTime:17258.8
;Volume:707.871
;Resin:
;Weight:708.201
;Price:35.41
;LayerHeight:0.1
;ResolutionX:3840
;ResolutionY:2400
;MachineX:192
;DisplayHeight:120
;MachineZ:220
;Mirror:0
;LayerCount:2118
;BottomLayerCount:4
;TransitionLayerCount:4
;BottomWaitTimeBeforeCure:3
;WaitTimeBeforeCure:2.5
;BottomExposureTime:3
;ExposureTime:3
;BottomWaitTimeAfterCure:3
;WaitTimeAfterCure:1
;BottomLiftHeight:2
;BottomLiftSpeed:200
;BottomRetractSpeed:400
;LiftHeight:5
;LiftSpeed:400
;RetractSpeed:400
;BottomWaitTimeAfterLift:0
;WaitTimeAfterLift:0
;BottomLightPWM:255
;PWMLight:255
;AntiAliasing:8
;BoundingRectanglePx:{X=197,Y=98,Width=3446,Height=2215}
;BoundingRectangleMm:{X=9.85,Y=9,Width=4.9,Height=4}

;START_GCODE_BEGIN
MSLA_DISPLAY_VALIDATE RESOLUTION=3840,2400 PIXEL=0.05,0.05
SET_PRINT_STATS_INFO TOTAL_LAYER=2118
PRINT_START LAYERS=2118 HEIGHT=211.8 BOUNDS=9.85,4.9,172.3,110.75 VOLUME=708565.94 MATERIALML=707.871
;END_GCODE_BEGIN

;LAYER_START:0
;PositionZ:0.1mm
SET_PRINT_STATS_INFO CURRENT_LAYER=1
BEFORE_LAYER_CHANGE LAYER=1 HEIGHT=0.1 PIXELS=1626474 BOUNDS=13.45,7,162,106.05 AREA=4066.185 VOLUME=406.619 MATERIALML=0.4066
M73 P0 R282.65;Set print progress
M6054 F"1.png";Show image
G0 Z2.1 F200;Z Lift (1)
G0 Z8.1 F400;Z Lift (2)
G0 Z0.1 F400;Retract (1)
G4 P3000;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P3000;Wait after cure
;LAYER_END

;LAYER_START:1
;PositionZ:0.2mm
SET_PRINT_STATS_INFO CURRENT_LAYER=2
BEFORE_LAYER_CHANGE LAYER=2 HEIGHT=0.2 PIXELS=1638068 BOUNDS=13.4,6.9,162.1,106.2 AREA=4095.17 VOLUME=409.517 MATERIALML=0.4095
M73 P0.05 R282.45;Set print progress
M6054 F"2.png";Show image
G0 Z2.2 F200;Z Lift (1)
G0 Z8.2 F400;Z Lift (2)
G0 Z0.2 F400;Retract (1)
G4 P3000;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P3000;Wait after cure
;LAYER_END

;LAYER_START:2
;PositionZ:0.3mm
SET_PRINT_STATS_INFO CURRENT_LAYER=3
BEFORE_LAYER_CHANGE LAYER=3 HEIGHT=0.3 PIXELS=531661 BOUNDS=14.9,8.4,159.15,103.25 AREA=1329.153 VOLUME=132.915 MATERIALML=0.1329
M73 P0.09 R282.26;Set print progress
M6054 F"3.png";Show image
G0 Z2.3 F200;Z Lift (1)
G0 Z8.3 F400;Z Lift (2)
G0 Z0.3 F400;Retract (1)
G4 P3000;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P3000;Wait after cure
;LAYER_END

;LAYER_START:3
;PositionZ:0.4mm
SET_PRINT_STATS_INFO CURRENT_LAYER=4
BEFORE_LAYER_CHANGE LAYER=4 HEIGHT=0.4 PIXELS=518615 BOUNDS=14.9,8.4,159.1,103.2 AREA=1296.538 VOLUME=129.654 MATERIALML=0.1297
M73 P0.14 R282.06;Set print progress
M6054 F"4.png";Show image
G0 Z2.4 F200;Z Lift (1)
G0 Z8.4 F400;Z Lift (2)
G0 Z0.4 F400;Retract (1)
G4 P3000;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P3000;Wait after cure
;LAYER_END

;LAYER_START:4
;PositionZ:0.5mm
SET_PRINT_STATS_INFO CURRENT_LAYER=5
BEFORE_LAYER_CHANGE LAYER=5 HEIGHT=0.5 PIXELS=460091 BOUNDS=15,8.5,158.9,103.05 AREA=1150.228 VOLUME=115.023 MATERIALML=0.115
M73 P0.19 R281.87;Set print progress
M6054 F"5.png";Show image
G0 Z5.5 F400;Z Lift (1)
G0 Z0.5 F400;Retract (1)
G4 P2500;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P1000;Wait after cure
;LAYER_END

;LAYER_START:5
;PositionZ:0.6mm
SET_PRINT_STATS_INFO CURRENT_LAYER=6
BEFORE_LAYER_CHANGE LAYER=6 HEIGHT=0.6 PIXELS=403906 BOUNDS=15.05,8.6,158.8,102.85 AREA=1009.765 VOLUME=100.977 MATERIALML=0.101
M73 P0.24 R281.73;Set print progress
M6054 F"6.png";Show image
G0 Z5.6 F400;Z Lift (1)
G0 Z0.6 F400;Retract (1)
G4 P2500;Wait before cure
M1400 S255 P3000;Exposure for 3s
G4 P1000;Wait after cure
;LAYER_END

...
...
...
;START_GCODE_END
PRINT_END
;END_GCODE_END
;<Completed>
; total layers count = 2118
```

**Comment by meteyou** [comment]
Your Gcode will not be parsed by Moonraker, because Moonraker doesn't know your Slicer. So, an estimated time or something similar will not be displayed because of this.

But your Layer number should work with the `SET_PRINT_STATS_INFO` command. I will double-check the Layer numbers.

To get the other metadata working, you can try changing the "generated" line in your gcode, which Moonraker thinks is a Prusaslicer gcode.

**Reply by sn4k3** [reply]
I added the following line copied from PS into my gcode:
`; generated by PrusaSlicer 2.7.4+win64 on 2024-04-29 at 14:31:17 UTC`

Under files it now show slicer, but all other fields are empty. I expect at least to populate layer height, print height and time.
![image](https://github.com/mainsail-crew/mainsail/assets/113281/a55053e1-63e3-4146-adc4-129cd9d53a47)

Printing the file does same behavior as before:

![image](https://github.com/mainsail-crew/mainsail/assets/113281/f99fe6f8-0013-4980-8736-ffe1741f0b0f)

However if I cancel print the total shows:

![image](https://github.com/mainsail-crew/mainsail/assets/113281/b0f9950a-6784-417f-88e5-494c917ba305)

It looks like for some reason the `SET_PRINT_STATS_INFO` is not taken into account.
It's also strange the ETA and estimate does not work. It have the data with `M73`, and I tried both File and Slicer options, none give an ETA.

**Comment by meteyou** [comment]
when you dont see any more informations, moonraker cannot find/identify it.

the estimated time will only start, when your job "really starts". i think klipper start it with the first extrusion, but i'm not 100% sure. pls double-check it yourself with the klipper code.

**Reply by sn4k3** [reply]
You right, I modified a line and it now shows the stats, however the layer count still freeze.

![image](https://github.com/mainsail-crew/mainsail/assets/113281/b1e9ad78-c5cb-44c8-89f5-cbd4fb33724d)

The mod (line highlight):

![image](https://github.com/mainsail-crew/mainsail/assets/113281/9b28c2f3-d03c-4253-8e73-189a81a0e1dd)

**Reply by sn4k3** [reply, marked as answer]
Found the problem with layer count. The command `SET_PRINT_STATS_INFO TOTAL_LAYER=x` must be set after home and not before. I was setting it as first line and that cause problem.
